File: Second Year/PDOO/Napakalaki/src/NapakalakiGame/BadConsequence.java
/*  
 *  Programación y diseño orientado a objetos
 *  Grado en Ingeniería Informática
 * 
 *  
 *
 *  Angela Castro Jiménez
 *  Francisco Carrillo Pérez
 */
package NapakalakiGame;

import java.util.ArrayList;


public class BadConsequence {
    
    private String text;
    private int levels;
    private int nVisibleTreasures;
    private int nHiddenTreasures;
    private boolean death;
    
    private ArrayList<TreasureKind> specificHiddenTreasures = new ArrayList();
    private ArrayList<TreasureKind> specificVisibleTreasures = new ArrayList();
    
    public BadConsequence(){
        text = "";
        levels = 0;
        nVisibleTreasures = 0;
        nHiddenTreasures = 0;
    }
    
    public BadConsequence (String t, int l, int nVisible, int nHidden){
        text = t;
        levels = l;
        nVisibleTreasures = nVisible;
        nHiddenTreasures = nHidden;
    }
    
    public BadConsequence (String t, int l, ArrayList<TreasureKind> v, ArrayList<TreasureKind> h){
        text = t;
        levels = l;
        specificVisibleTreasures = v;
        specificHiddenTreasures = h;
    }
    
    public BadConsequence (String t, boolean mideath){
        text = t;
        death = mideath;
    }
    
    
    public String getText(){
        return text;
    }  
    
    public int  getLevels(){
         return levels;
    }
    
    public int  getnVisibleTreasures(){
         return nVisibleTreasures;
    }
    
    public int  getnHiddenTreasures(){
         return nHiddenTreasures;
    }
     
    public boolean getDeath(){
         return death;
    }
     
     
     public ArrayList<TreasureKind>  getSpecificHiddenTreasures(){
         return specificHiddenTreasures;
     }
     
     
      public ArrayList<TreasureKind>  getSpecificVisibleTreasures(){
         return specificVisibleTreasures;
      }
     
    boolean myBadConsequenceIsDeath(){
         return death;
    }
     
    public boolean isEmpty(){
          return (nVisibleTreasures == 0 && nHiddenTreasures == 0 && 
            (specificHiddenTreasures == null || specificHiddenTreasures.isEmpty()) && (specificVisibleTreasures == null || specificVisibleTreasures.isEmpty()) );  
    }
    
    //Actualiza el mal rollo que tiene que cumplir el jugador, en función del tipo de tesoro de t y
    //del tipo de mal rollo que tenga que cumplir el jugador.

    public void substractVisibleTreasure(Treasure t){
        //no tengo claro que sea así
        specificVisibleTreasures.remove(t.getType());
    }
    
    
     public void substractHiddenTreasure(Treasure t){
         specificHiddenTreasures.remove(t.getType());
     }
    
    
     public BadConsequence adjustToFitTreasureLists( ArrayList<Treasure> v, ArrayList<Treasure> h){
         BadConsequence bc;
         if(this.specificHiddenTreasures == null && this.specificVisibleTreasures == null){
             bc=new BadConsequence("",0,this.nVisibleTreasures, this.nHiddenTreasures);
             if(bc.nHiddenTreasures>h.size())
                   bc.nHiddenTreasures = h.size();
             if(bc.nVisibleTreasures> v.size())
                    bc.nVisibleTreasures = v.size();
         }
         else{
            ArrayList<TreasureKind> tVisible = new ArrayList();
            ArrayList<TreasureKind> tHidden = new ArrayList();

           
           for (TreasureKind t: this.specificHiddenTreasures) {
               int cont = 0;
               boolean enc =false;
               while (!enc && cont<h.size()){
                   if(t==h.get(cont).getType())
                       enc = true;
                   cont++;
               }
               if (enc)
                   tHidden.add(t);
           }
           for (TreasureKind t: this.specificVisibleTreasures) {
               int cont = 0;
               boolean enc =false;
               while (!enc && cont<v.size()){
                   if(t==v.get(cont).getType())
                       enc = true;
                   cont++;
               }
               if (enc)
                   tVisible.add(t);
           }
           

         bc = new BadConsequence("", 0, tVisible, tHidden);
         }
           return bc;
     }
     
     public String toString(){
        return "Mal Rollo: "+text
                +" \nPierdes " + Integer.toString(levels)+" niveles."
                +"\nNumero de tesoros visibles perdidos: "+ Integer.toString(nVisibleTreasures)+"."
                +"\nNumero de tesoros ocultos perdidos: "+ Integer.toString(nHiddenTreasures)+"."
                +"\nMuerte: "+ Boolean.toString(death);
     }
     

     
      /* PRUEBA 
        public static void main(String args[]){
           BadConsequence malrollo = new BadConsequence ("Pierdes un turno",2,3,2);
           System.out.println(malrollo.getText());
           System.out.println(malrollo.getLevelsPerdidos());
           System.out.println(malrollo.getnVisiblePerdidos()); 
           System.out.println(malrollo.getnHiddenPerdidos());
           System.out.println(malrollo.toString());
        } */
   
}
    
    
    

File: Second Year/PDOO/Napakalakirb/lib/BadConsequence.rb
# encoding: UTF-8
#
#   Programación y diseño orientado a objetos
#   Grado en Ingeniería Informática
#
#  Ángela Castro Jimenez
#  Francisco Carrillo Pérez
module Napakalaki
  class BadConsequence
    attr_accessor :text, :levels, :nVisibleTreasures, :nHiddenTreasures, :death,:specificHiddenTreasures,:specificVisibleTreasures
    
   #el metodo new no puede ser llamado fuera de un constructor
    private_class_method :new
    def initialize(text,levels,nVisible,nHidden,death,specificHiddenTreasures,specificVisibleTreasures)
      @text = text
      @levels = levels
      @nVisibleTreasures = nVisible
      @nHiddenTreasures = nHidden
      @death = death
      @specificHiddenTreasures = specificHiddenTreasures
      @specificVisibleTreasures = specificVisibleTreasures
    end

    def self.newDeath(t)  
        new(t, 0, 0, 0,true, [], [])
    end

    def self.newLevelNumberOfTreasures(t,l,nVisible,nHidden)
      new(t,l,nVisible,nHidden, [],[],false)
    end
    
    def self.BadConsequence(t,l,v,h)
      new(t,l,nil,nil,nil,v,h)
    end
     def BadConsequence.newLevelSpecificTreasures(text, levels, specificVisibleTreasures, specificHiddenTreasures)
      new(text,levels,0,0,specificVisibleTreasures,specificHiddenTreasures,false)
    end
    
    
    def isEmpty()
      return (nVisibleTreasures == 0 && nHiddenTreasures == 0 && levels == 0 )
    end
  

       def getSpecificVisibleTreasures()
          return @specificVisibleTreasures
       end   
    
      def getSpecificHiddenTreasures()
         return @specificHiddenTreasures
       end
    
  
        def substractVisibleTreasure(t)
           tipo=t.type
          encontrado=false
          for i in (0..@specificHiddenTreasures.size-1)
             break if (encontrado)
            if(@specificHiddenTreasures.at(0)==tipo.at(0))
              @specificHiddenTreasures.delete_at(i)
              encontrado=true
            end
          end
          @nHiddenTreasures=@nHiddenTreasures-1 if(@nHiddenTreasures>0)  
        end
    
    
        def substractHiddenTreasure(t)
           tipo=t.type
           encontrado=false
           for i in (0..@specificHiddenTreasures.size-1)
             break if (encontrado)
              if(@specificHiddenTreasures.at(0)==tipo.at(0))
               @specificHiddenTreasures.delete_at(i)
               encontrado=true
              end
            end
            @nHiddenTreasures=@nHiddenTreasures-1 if(@nHiddenTreasures>0) 
        end
   
    
    def adjustToFitTreasureLists( visible,  hidden)
        encontrado = false
      nh=0
      nv=0
      antiguaBadConsequence = Napakalaki.instance.getCurrentMonster.getBadConsequence
      
      if(antiguaBadConsequence.getNVisibleTreasures > 0 && visible.size > 0 )  
        if(antiguaBadConsequence.getNVisibleTreasures >= visible.size)
          nv=visible.size
        else
          nv=antiguaBadConsequence.getNVisibleTreasures
        end
      end
      
      if(antiguaBadConsequence.getNHiddenTreasures > 0 && hidden.size > 0 )  
        if(antiguaBadConsequence.getNHiddenTreasures >= hidden.size)
          nh=hidden.size
        else
          nh=antiguaBadConsequence.getNHiddenTreasures
        end
      end
      
      nuevaBadConsequence = BadConsequence.newLevelNumberOfTreasures("", 0, nv, nh)
      
      if(antiguaBadConsequence.specificVisibleTreasures.size > 0 || antiguaBadConsequence.specificHiddenTreasures.size > 0)
        for i in (0..visible.size-1)
          encontrado=false
          for j in (0..antiguaBadConsequence.specificVisibleTreasures.size-1)
            break if (encontrado)
            tip1=visible.at(i).type
            tip2=antiguaBadConsequence.specificVisibleTreasures.at(j)
            if(tip1.at(0)==(tip2))
              nuevaBadConsequence.specificVisibleTreasures.push(antiguaBadConsequence.specificVisibleTreasures.at(j))
              antiguaBadConsequence.specificVisibleTreasures.delete_at(j)
              encontrado=true
            end
          end
        end
        
        for i in (0..hidden.size-1)
          encontrado=false
          for j in (0..antiguaBadConsequence.specificHiddenTreasures.size-1)
            break if (encontrado)
            tip1=hidden.at(i).type
            tip2=antiguaBadConsequence.specificHiddenTreasures.at(j)
            if(tip1.at(0)==tip2)
              nuevaBadConsequence.specificHiddenTreasures.push(antiguaBadConsequence.specificHiddenTreasures.at(j))
              antiguaBadConsequence.specificHiddenTreasures.delete_at(j)
              encontrado=true
            end
          end
        end
      end
      
      return nuevaBadConsequence
    end

    def myBadConsequenceIsDeath
        return @death
    end
    def getdeath
      @death
    end

    def to_s 
      "Mal rollo: #{@text} \n Niveles: #{@levels} \n Num. Tesoros Visibles: #{@nVisibleTreasures}
        \n Num. Tesoros Ocultos: #{@nHiddenTreasures} \n Muerte: #{@death}"
    end

  end
end